### ISSUE TITLE
Backup temporary directory not unmounted correctly

### ISSUE THREAD

**Issue by tofbouf** [YunoHost/issues#1809]
Hi.

I have linked /home/yunohost.backup to some extern HDD (to not harras the raspberry Pi SD card):

lrwxrwxrwx   1 root      root        29 Oct 24  2020 yunohost.backup -> /mnt/extern1/backups/yunohost

When a backup needs some mounting/binding in the backup temporary dir (let's say backing up roundcube), it binds
  '/mnt/extern1/var/www/roundcube'
onto
  '**/home/yunohost.backup**/tmp/auto_roundcube/apps/roundcube/backup/mnt/extern1/var/www/roundcube'

But, in /etc/mtab, this is 'real' path that appears:

/dev/sda1 **/mnt/extern1/backups/yunohost**/tmp/auto_roundcube/apps/roundcube/backup/mnt/extern1/var/www/roundcube ext4 ro,noatime 0 0

After the backup is done, the clean() method tries to unmount anything starting with **/home/yunohost.backup**/tmp/auto_roundcube, and of course it found no such mounting point.

On the next backup, the temporary directory is detected existing, and the recursive unmount procedure is also trying to unmount /home/yunohost.backup/tmp/auto_roundcube, buit it fails in the same way.
The backup temporary directory cannot be cleaned, and the backup fails.

To 'fix' the issues, it needs either to reboot (the mounting points will be canceled) or to unmount manually the bindings.

In backup.py, in the recursive_umount function, it would need to 'translate' directory into the same way it appears in /etc/mtab, that is to say translating any symbolic link into its target, or something like this.

Note that I suck in Pyhton, and I have not a single idea how to do this...

**Comment by tofbouf** [comment]
To translate the path, **realpath** or **readlink -f** will do the job:
```
$/home/admin# realpath /home/yunohost.backup/tmp/auto_roundcube/
/mnt/extern1/backups/yunohost/tmp/auto_roundcube
$/home/admin# readlink -f /home/yunohost.backup/tmp/auto_roundcube/
/mnt/extern1/backups/yunohost/tmp/auto_roundcube
```

**Comment by zamentur** [comment]
On the same code there is this bug :  https://github.com/YunoHost/issues/issues/1767
This ticket seems not to be a duplicate but that's in the same zone.

**Comment by zamentur** [comment]
Unrelated: for performance reasons you shouldn't symlink /home/yunohost.backup but /home/yunohost.backup/archives (cause some operation could do `mv` from /home/yunohost.backup/tmp, so it's preferable to have this on same disk...)

**Comment by tofbouf** [comment]
> Unrelated: for performance reasons you shouldn't symlink /home/yunohost.backup but /home/yunohost.backup/archives (cause some operation could do `mv` from /home/yunohost.backup/tmp, so it's preferable to have this on same disk...)

In fact, the symlinking of /home/yunohost.backup was completely useless, as /home is already symlinked to the external HDD (/mnt/extern1/home)... So, removing the symlink of /home/yunohost.backup did not help...

**Comment by tofbouf** [comment]
I've made a test, modifiying _recursive_umount() in backup.py:

line 2661:
`        if len(line) >= 3 and line.split(" ")[2].startswith(directory)`
into
`        if len(line) >= 3 and line.split(" ")[2].startswith(os.path.realpath(directory))`

It is now correctly unmounting any remaining bindings either at the start of backup (old bindings from fail backup) and at the end of backup when it is finished OK.

Tested on a apps + data +conf backup, both with local backup and borg backup.

I can made a pull request if needed.
